.github/workflows/ci.yaml

```diff
@@ -1,34 +1,161 @@
 name: CI build
 
+concurrency:
+  group: ci-${{ github.run_id }}
+  cancel-in-progress: true
+
 on:
-  push:
-    branches:
-      - main
-      - v[0-9]+
-      - v[0-9]+.[0-9]+
-      - cryostat-v[0-9]+.[0-9]+
-  pull_request:
+  pull_request_target:
+    types:
+      - opened
+      - reopened
+      - synchronize
+      - labeled
+      - unlabeled
     branches:
       - main
       - v[0-9]+
       - v[0-9]+.[0-9]+
       - cryostat-v[0-9]+.[0-9]+
 
 jobs:
-  build:
+  check-before-build:
+    runs-on: ubuntu-latest
+    steps:
+    - name: Fail if safe-to-test is not applied
+      if: github.repository_owner == 'cryostatio' && (!contains(github.event.pull_request.labels.*.name, 'safe-to-test'))
+      run: exit 1
 
+  build-core:
     runs-on: ubuntu-latest
+    outputs:
+      core_version: ${{ steps.get_core_version.outputs.core_version }}
+    steps:
+    - uses: actions/checkout@v4
+    - id: get_core_version
+      run: |
+        echo "core_version=$(mvn help:evaluate -Dexpression=project.version -q -DforceStdout)" >> $GITHUB_OUTPUT
+        echo "name=$(mvn help:evaluate -Dexpression=project.artifactId -q -DforceStdout)" >> $GITHUB_OUTPUT
+    - uses: actions/setup-java@v3
+      with:
+        java-version: '17'
+        distribution: 'temurin'
+    - uses: actions/cache@v3
+      with:
+        path: ~/.m2
```

Comment:
I think this is what I was missing @andrewazores the other day you mentioned.

Reply:
oh yea Thuan and I spent a long time trying to figure out how to successfully cache the maven files haha

Reply:
Looks good to me.